src/resources/tls.cc

```diff
@@ -99,9 +100,73 @@ int BaseMbedTlsSocket::add_root_certificate(X509Certificate* cert) {
   return 0;
 }
 
-void BaseMbedTlsSocket::apply_certs() {
+uint32 BaseMbedTlsSocket::hash_subject(uint8* buffer, int length) {
+  // Matching should be case independent for ASCII strings, so lets just zap
+  // all the 0x20 bits, since we are just doing a fuzzy match.
+  for (int i = 0; i < length; i++) buffer[i] |= 0x20;
+  return Utils::crc32(0xce77509, buffer, length);
+}
+
+// Use the unparsed certificates on the process to find the right one
+// for this connection.
+static int toit_tls_find_root(void* context, const mbedtls_x509_crt* certificate, mbedtls_x509_crt** chain) {
+  BaseMbedTlsSocket* socket = unvoid_cast<BaseMbedTlsSocket*>(context);
+  Process* process = socket->resource_group()->process();
+
+  uint8 issuer_buffer[MAX_SUBJECT];
+  int ret = mbedtls_x509_dn_gets(char_cast(&issuer_buffer[0]), MAX_SUBJECT, &certificate->issuer);
+  if (ret < 0) goto failed;
+  if (ret >= MAX_SUBJECT) {
+    ret = MBEDTLS_ERR_ASN1_BUF_TOO_SMALL;
+    goto failed;
+  } else {
+    uint32 issuer_hash = BaseMbedTlsSocket::hash_subject(issuer_buffer, ret);
+
+    *chain = null;
+    mbedtls_x509_crt cert;
+    mbedtls_x509_crt_init(&cert);
+    mbedtls_x509_crt** last = chain;
+    bool found_root_with_matching_subject = false;
+    Locker locker(OS::scheduler_mutex());
+    for (auto unparsed : process->root_certificates(locker)) {
+      if (unparsed->subject_hash() != issuer_hash) continue;
+      mbedtls_x509_crt* cert = _new mbedtls_x509_crt;
+      if (!cert) {
+        ret = MBEDTLS_ERR_X509_ALLOC_FAILED;
+        goto failed;
+      }
+
+      mbedtls_x509_crt_init(cert);
+      if (X509ResourceGroup::is_pem_format(unparsed->data(), unparsed->length())) {
+        ret = mbedtls_x509_crt_parse(cert, unparsed->data(), unparsed->length());
+      } else {
+        ret = mbedtls_x509_crt_parse_der_nocopy(cert, unparsed->data(), unparsed->length());
+      }
+      if (ret != 0) goto failed;
+      found_root_with_matching_subject = true;
```

Comment:
Do we need to continue to run through the rest of the unparsed certicates? Use `break` to get out quickly?

If you cannot, there should probably be a comment here that explains why we continue the hunt.